qiskit/pulse/schedule.py

```diff
@@ -1222,24 +1490,173 @@ def assign_parameters(
+
+        # Update parameters in the current scope
+        self._parameter_manager.assign_parameters(pulse_program=self, value_dict=value_dict)
+
+        for subroutine in self._reference_manager.values():
+            # Also assigning parameters to the references associated with self.
+            # Note that references are always stored in the root program.
+            # So calling assign_parameters from nested block doesn't update references.
+            if subroutine is None:
+                continue
+            subroutine.assign_parameters(value_dict=value_dict, inplace=True)
+
+        return self
+
+    def assign_references(
+        self,
+        subroutine_dict: Dict[Union[str, Tuple[str, ...]], "ScheduleBlock"],
+        inplace: bool = True,
+    ) -> "ScheduleBlock":
+        """Assign schedules to references.
+
+        It is only capable of assigning a schedule block to immediate references
+        which are directly referred within the current scope.
+        Let's see following example:
+
+        .. code-block:: python
+
+            from qiskit import pulse
+
+            with pulse.build() as subroutine:
+                pulse.delay(10, pulse.DriveChannel(0))
+
+            with pulse.build() as sub_prog:
+                pulse.reference("A")
+
+            with pulse.build() as main_prog:
+                pulse.reference("B")
+
+        In above example, the ``main_prog`` can refer to the subroutine "root::B" and the
+        reference of "B" to program "A", i.e., "B::A", is not defined in the root namespace.
+        This prevents breaking the reference "root::B::A" by the assignment of "root::B".
+        For example, if a user could indirectly assign "root::B::A" from the root program,
+        one can later assign another program to "root::B" that doesn't contain "A" within it.
+        In this situation, a reference "root::B::A" would still live in
+        the reference manager of the root.
+        However, the subroutine "root::B::A" would no longer be used in the actual pulse program.
+        To assign subroutine "A" to ``nested_prog`` as a nested subprogram of ``main_prog``,
+        you must first assign "A" of the ``sub_prog``,
+        and then assign the ``sub_prog`` to the ``main_prog``.
+
+        .. code-block:: python
+
+            sub_prog.assign_references({("A", ): nested_prog}, inplace=True)
+            main_prog.assign_references({("B", ): sub_prog}, inplace=True)
+
+        Alternatively, you can also write
+
+        .. code-block:: python
+
+            main_prog.assign_references({("B", ): sub_prog}, inplace=True)
+            main_prog.references[("B", )].assign_references({"A": nested_prog}, inplace=True)
+
+        Here :attr:`.references` returns a dict-like object, and you can
+        mutably update the nested reference of the particular subroutine.
+
+        .. note::
+
+            Assigned programs are deep-copied to prevent an unexpected update.
```

Comment:
What is the concern here?

Reply:
This is to make the principle clear. At the level of schedule, a key and referenced object should be unique mapping. So even if the source is the same object, it will be independent when assigned to the different keys. Thus, parameter link (mapto) should be implemented at the level of the calibrations framework.